contracts/stake/src/state.rs

```diff
@@ -13,8 +13,81 @@ use dusk_bls12_381_sign::PublicKey;
 use dusk_bytes::Serializable;
 use rusk_abi::State;
 
-use stake_contract_types::*;
-use transfer_contract_types::*;
+use phoenix_core::transaction::*;
+
+type BlockHeight = u64;
+
+/// Maturity of the stake
+const MATURITY: u64 = 2 * EPOCH;
+
+/// Epoch used for stake operations
+const EPOCH: u64 = 2160;
+
+#[derive(Debug, Clone)]
+pub struct StakeDataWrapper(pub StakeData);
+
+impl StakeDataWrapper {
+    /// Returns the value of the reward.
+    #[must_use]
+    pub const fn reward(&self) -> u64 {
+        self.0.reward
+    }
+
+    /// Returns the interaction count of the stake.
+    #[must_use]
+    pub const fn counter(&self) -> u64 {
+        self.0.counter
+    }
+
+    /// Insert a stake [`amount`] with a particular `value`, starting from a
+    /// particular `block_height`.
+    ///
+    /// # Panics
+    /// If the value is zero or the stake already contains an amount.
+    pub fn insert_amount(&mut self, value: u64, block_height: BlockHeight) {
+        assert_ne!(value, 0, "A stake can't have zero value");
+        assert!(
+            self.0.amount.is_none(),
+            "Can't stake twice for the same key!"
+        );
+
+        let eligibility = Self::eligibility_from_height(block_height);
+        self.0.amount = Some((value, eligibility));
+    }
+
+    /// Increases the held reward by the given `value`.
+    pub fn increase_reward(&mut self, value: u64) {
+        self.0.reward += value;
+    }
+
+    /// Removes the total [`amount`] staked.
+    ///
+    /// # Panics
+    /// If the stake has no amount.
+    pub fn remove_amount(&mut self) -> (u64, BlockHeight) {
+        self.0
+            .amount
+            .take()
+            .expect("Can't withdraw non-existing amount!")
+    }
+
+    /// Sets the reward to zero.
+    pub fn deplete_reward(&mut self) {
+        self.0.reward = 0;
+    }
+
+    /// Increment the interaction [`counter`].
+    pub fn increment_counter(&mut self) {
+        self.0.counter += 1;
+    }
+
+    /// Compute the eligibility of a stake from the starting block height.
```

Comment:
Leaving a reminder to update the description: it should explain that "maturation period" starts from the end of the current epoch.